src/core/or/circuitpadding.c

```diff
@@ -1199,6 +1257,17 @@ circpad_machine_transitioned_to_end(circpad_machineinfo_t *mi)
 {
   const circpad_machine_t *machine = CIRCPAD_GET_MACHINE(mi);
 
+  /** If this machine handles cirucit lifetime on its own, shut the circuit
+   *  down now if we must.
+   *
+   *  XXX We ignore should_negotiate_end in this case */
+  if (machine->manage_circ_lifetime) {
+    circuit_t *on_circ = mi->on_circ;
+    if (mi->circuit_was_asked_to_be_closed) {
+      circuit_mark_for_close(on_circ, END_CIRC_REASON_NONE);
+    }
```

Comment:
The XXX says "We ignore should_negotiate_end in this case" but we don't return here. I kinda think we should return here and not negotiate the end. Otherwise, circuits torn down due to circpad can have an end cell sent before they are closed, which seems like a (minor) information leak.

Reply:
Ye I should return here. That's a bug.